core/src/main/scala/scalaz/IList.scala

```diff
@@ -240,6 +240,9 @@ sealed abstract class IList[A] extends Product with Serializable {
   def map[B](f: A => B): IList[B] =
     foldRight(IList.empty[B])(f(_) :: _)
 
+  def foreach(f: A => Unit): Unit =
+    toList.foreach(f)
```

Comment:
My instinct here would be to write out the tail-recursive method to avoid constructing the `List` (and avoid relying on other methods for evaluation order).